### DISCUSSION TITLE
Middleware from Node Express Admins does not work , Rules only work if everyone.

### DISCUSSION THREAD

**Discussion by hattorihanzo86** [pocketbase/pocketbase#2908] [Q&A]
Hi there i created a middleware authenticate in node express

```js
const authenticate = async (req, res, next) => {
  const { userEmail, userPassword } = req.session;
  try {
    if (!userEmail || !userPassword) {
      throw new Error('Email or password not found in session');
    }
    const userData = await pb.admins.authWithPassword(userEmail, userPassword);
    req.user = userData;
    next();
  } catch (error) {
    res.status(401).json({ error: 'Authentication failed' });
  }
};
```

**and i can call it like this router**
```js
router.get('/list', authenticate, async (req, res) => {
   console.log('Getting list of compliance ...')
    try {
      const records = await pb.collection('compliance').getFullList({
        sort: '-created',
    });
    console.log(records)
       
    } catch (error) {
    res.status(500).json({ error: 'Failed getting list from collection ' + collection_name  + '...' });
    }
});
```

It really works but the problem is if you are admin you cant create/list/or anything in the crud if you change the rules
to everyone it works ? is there anything i could do because i dont want to make it like logging in before do that

**this work as admin but not your typically middleware**
```js
const loginGetCompliance = async (req, res) => {
  const { email, password } = req.body;
  //console.log('Logging in user ' + email + " ...")
  try {
    const userData = await pb.admins.authWithPassword(email, password);

    req.session.userEmail = email;
    req.session.userPassword = password;

    //console.log(pb);

    const records = await pb.collection('compliance').getFullList({
      sort: '-created',
  });

    console.log(records)


    res.json({ message: 'Login successful', userData });
    //console.log(userData)
  } catch (error) {
    res.status(401).json({ error: 'Login failed' });
  }
};
```

**Comment by ganigeorgiev** [comment]
> It really works but the problem is if you are admin you cant create/list/or anything in the crud if you change the rules
to everyone it works ? is there anything i could do because i dont want to make it like logging in before do that

I don't understand what do you mean by this. API rules don't apply for admins. If you have successfully authenticated as admin, then the admin account should have access to everything.

Are you trying to authenticate as regular app user? If Yes, then you should use `pb.collection("users").authWithPassword(email, password)` (_`users` in this case is the name of your auth collection_). If you haven't already, you can also check https://pocketbase.io/docs/authentication/.

**Reply by normanmadrid** [reply]
Hi gani let me clarify for my colleague,

we have two async functions, a login function where we call:
authenticate = async (req, res, next) => {
    const userData = await pb.admins.authWithPassword(email, password);
}

and a listing function where we call:
router.get('/list', authenticate, async (req, res) => {
    const records = await pb.collection('compliance').getFullList({
      sort: '-created',
    })
}
  
when called in two separate functions, the listing function returns a 'only admins can perform this action'
but when called in the same function, it recognizes us as admin, which is what's ideal in our implementation, we want to call these functions as admin. Pocketbase is awesome and appreciate all the help, thanks!

**Reply by ganigeorgiev** [reply]
I'm sorry, I still don't understand the described issue. How the PocketBase JS SDK is initialized? Are you using a single global instance or create a new SDK client for each request? Could you create a **minimal** reproducible repo?

**Reply by normanmadrid** [reply]
https://github.com/DalmoreIT/pb-api
here ya go

**Reply by normanmadrid** [reply]
To further elaborate, we are trying to implement it so that we only have to authenticate as admin once in its own function, and to have our pocketbase instance recognize us as admin when we call getFullList in a separate async  function . As it stands if 'admins.authWithPassword' and 'getFullList' are not in the same function, our call gets the response 'Only admins can perform this action'.

**Reply by ganigeorgiev** [reply]
It seems that you are creating a new PocketBase client instance in each file, for example https://github.com/DalmoreIT/pb-api/blob/main/sources/collections/companies.mjs#L6.

When calling `new PocketBase()` in a node env you are creating a new isolated AuthStore with each instance, meaning that the auth state loaded from https://github.com/DalmoreIT/pb-api/blob/main/sources/auth.mjs is not persisted and will be different than the one from the files in your collections directory.

I don't really know what to recommend and how to explain it better as what you are trying to do is not something that can be accomplished easily.

I don't understand why do you want to be able to authenticate in your node application as different PocketBase admins, but if you want to do this, then you'll need a different structure similar to the [SSR guide](https://github.com/pocketbase/js-sdk#ssr-integration), aka. you need to create a new PocketBase instance for each request and keep the created instance in a local scoped request variable. The authState between different requests could be loaded by storing the token in a cookie or some other express session mechanism. Also note that calling `authWithPassword()` for each request is not recommended as it is slow due to the bcrypt hash generation and it is recommended to keep the generated token in your session object instead of the plain password in https://github.com/DalmoreIT/pb-api/blob/main/sources/auth.mjs#L36.

**Reply by hattorihanzo86** [reply]
Hi @ganigeorgiev  thanks so much in short we created a authenticated middleware that use admin once it login it's authenticated for every collection access. I try also the token to be on the header actually it really works but some wierd problem the Admin is not distinguished in the api calls. 


Scenario 1:
When we call this it know's that it is admin
```
const userData = await pb.admins.authWithPassword(userEmail, userPassword);
const records = await pb.collection('companies').getFullList({ sort: '-created' });
console.log(records);
```

Scenario 2: 
We create a custom middleware authenticate from this it should be authenticated and
this proven the middleware works we expected but seems the scenario 1 always work to get the list.
```
router.get('/company/list', authenticate, async (req, res) => {
  try {
    const records = await pb.collection('companies').getFullList({ sort: '-created' });
    res.json(records);
  } catch (error) {
    res.status(500).json({ error: 'Company getting list from server failed!' });
  }
});
```

We are very grateful that we found PocketBase and we are trying to explore more.
Hope you can help us.

**Reply by ganigeorgiev** [reply]
@hattorihanzo86 I'm sorry, I don't think I can help you as I really don't know how to explain it better.

You deleted your sample repo, but the issue why the `authenticate` middleware doesn't work is explained in the comment above https://github.com/pocketbase/pocketbase/discussions/2908#discussioncomment-6433683. The `pb` instance variable inside your route is different then the `pb` that is created in the middleware. If you place the `router.get(...)` call inside the same file where the middleware is and the `new PocketBase` is located it will work, but again, please read the above comment as this is not correct since as far as I understand you want to enable multiple admins to being able to authenticate and this will fail because requests from different admin accounts will write over the same `pb.authStore`. This is not something PocketBase specific. Keep in mind that the Node.js execution is usually one long running process and the lifecycle of the variables is not the same as in PHP or in a lambda/cloud-function context. It is OK to have a single global `pb` in Node.js **only** if you want to have a single PocketBase admin and there is no risk requests from different users to change the same `pb.authStore`.

**Reply by ganigeorgiev** [reply]
@hattorihanzo86, @normanmadrid This is the simple express example setup I can think of using a single PocketBase admin (the credentials could be loaded from env variables):

1. Create a single `pb.js` file that will keep the global PocketBase admin instance:
    ```js
    // pb.js
    import PocketBase from "pocketbase"

    const pb = new PocketBase("https://127.0.0.1:8090")

    export default pb
    ```

2. Create `middlewares.js` file that will contain your app middlewares:
    ```js
    import { isTokenExpired } from "pocketbase"
    import pb from "./pb.js"

    // note: could be loaded from env variables
    const ADMIN_EMAIL = "test@example.com"
    const ADMIN_PASS = "1234567890"

    // basic auth middleware to ensure that the global PocketBase client is authenticated.
    export async function auth(req, res, next) {
      if (pb.authStore.isValid) {
        // just refresh the token if it is going to expire in the next 30 minutes
        if (isTokenExpired(pb.authStore.token, 1800)) {
          await pb.admins.authRefresh()
        }
      } else {
        // authenticate with email/password
        await pb.admins.authWithPassword(ADMIN_EMAIL, ADMIN_PASS)
      }
      
      next()
    }
    ```

3. Create your app entrypoint file, eg. `index.js`:
    ```js
    // index.js
    import express  from "express"
    import pb       from "./pb.js"
    import { auth } from "./middlewares.js"

    const app = express()
    const port = 3000

    // you can register the auth middleware globally
    app.use(auth)

    // or per handler
    app.get('/', auth, async (req, res) => {
      // perform your PocketBase requests as usual
      // (the pb in this case is authenticated as admin via the middleware)
      const items = await pb.collection("example").getFullList()

      res.send('Hello World!')
    })

    app.listen(port, () => {
        console.log(`Example app listening on port ${port}`)
    })
    ```

4. Start the app using `node index.js` and navigate to `127.0.0.1:3000`.

_Again, the above is just an example for using a global `pb` instance. As mentioned earlier you can also attach the `pb` to the [`res.locals`](https://expressjs.com/en/5x/api.html#res.locals) prop._

_**Edit:** Changed the auth middleware to call `authRefresh` only if the token is going to expire in the next 30 minutes_

**Reply by normanmadrid** [reply]
This is too much, you are a saint gani! will test it out asap!

**Reply by hattorihanzo86** [reply]
Hi @ganigeorgiev  seems the middleware that you give is fine but the problem the api that where creating is prone to access without authentication first. So what i did is when the admin login i store there email , password into session.


middleware.mjs
```
export async function auth(req, res, next) {
  const { userEmail, userPassword } = req.session;
  if (pb.authStore.isValid) {
    // just refresh the token if it is going to expire in the next 30 minutes
    if (isTokenExpired(pb.authStore.token, 1800)) {
      await pb.admins.authRefresh()
    }
  } else {
    // authenticate with email/password
    await pb.admins.authWithPassword(userEmail, userPassword)
  }
  next()
}
```

Also i notice in this code it gives infinite loops it will display the result but in postman progress it will be infinite 
```
app.get('/', auth, async (req, res) => {
  // perform your PocketBase requests as usual
  // (the pb in this case is authenticated as admin via the middleware)
  const items = await pb.collection("example").getFullList()

  res.send('Hello World!')
})
```

Here is the solution and it works.
```
router.get('/list', auth, async (req, res) => {
  console.log('Getting list of compliance ...');
  try {
    const timeout = 5000; // Timeout in milliseconds (adjust as needed)
    const recordsPromise = pb.collection(collection_name).getFullList({
      sort: '-created',
    });
    const records = await Promise.race([
      recordsPromise,
      new Promise((resolve, reject) => setTimeout(reject, timeout))
    ]);
    console.log(records);
    res.json(records);
  } catch (error) {
    res.status(500).json({ error: 'Failed getting list from collection ' + collection_name + '...' });
  }
});
```


Thank you so much for your help we solve the problem and we established NodeJS PocketBased integration
successfully. We can't find this in the documents and the community hopes that with this experience others
can help them too.

**Reply by ganigeorgiev** [reply]
> Hi @ganigeorgiev seems the middleware that you give is fine but the problem the api that where creating is prone to access without authentication first.

The admin email and password are expected to be loaded from the `ADMIN_EMAIL` and `ADMIN_PASS` variables (eg. usually loaded from a config or as ENV variables).

Just to clarify to prevent some eventual misunderstanding, my example in https://github.com/pocketbase/pocketbase/discussions/2908#discussioncomment-6435951 is suitable **only** if you want your node application to use a single admin to comminicate with PocketBase. If you want to allow multiple admins to login via your Node.js app then the above will not work for the reasons explained in https://github.com/pocketbase/pocketbase/discussions/2908#discussioncomment-6433683 and https://github.com/pocketbase/pocketbase/discussions/2908#discussioncomment-6435488.

**Reply by hattorihanzo86** [reply]
hi there yes 1 admin only it require to login first because if we put the credentials in .env the nodeexpress server api that we made is accessible by anyone else. Thanks for the help :)

**Reply by ganigeorgiev** [reply]
> if we put the credentials in .env the nodeexpress server api that we made is accessible by anyone else. 

I'm not sure that I understand what you mean by this, but I'll not pursue it further.

If you've found a working solution - great, just make sure that you understand your app lifecycle and test your app security thoroughly because PocketBase admins have access to everything (including deleting entire collections).

**Reply by hattorihanzo86** [reply]
Yes thanks so much ... We will observe it...
